File: 24_高级数据结构/good java/problem/NumberSumLeqMatch.java
package template.problem;

/**
 * There are two row of numbers, each row contains 1, ...,n.
 * <br>
 * Given m, for row 1 number x and row 2 number y, if x + y <= m, there is an edge (x,y) between them.
 * <br>
 * You're asked to find a largest matching on this bipartite graph.
 */
public class NumberSumLeqMatch {
    long n;
    long m;
    long used;
    long type1;

    public NumberSumLeqMatch(long n, long m) {
        this.n = n;
        this.m = m;
        used = Math.min(n, m - 1);
        type1 = m - used - 1;
    }

    /**
     * Get the maximum matching
     */
    public long maxMatching() {
        return used;
    }

    /**
     * Get the partner of x, if not exist, return -1
     */
    public long partner(long x) {
        if (x > used) {
            return -1;
        }
        if (x <= type1) {
            return x;
        }
        return m - x;
    }

    /**
     * Get the maximum matching if remove a from row 1 and remove b from row 2
     */
    public long maxMatching(long a, long b) {
        if (a >= m && b >= m) {
            return used;
        }
        if (a + b < m && type1 <= 1) {
            return used - 2;
        }
        return used - 1;
    }
}

File: 24_高级数据结构/good java/NumberSumLeqMatch/NumberSumLeqMatch.go
package main

// 给定两个数组，每个数组包含1到n的数字。
// 给定m，对于数组1中的数字x和数组2中的数字y，如果x+y<=m，则它们之间有一条边(x,y)。
// 你需要找到这个二部图上的最大匹配。
type NumberSumLeqMatch struct {
	n, m        int
	used, type1 int
}

func NewNumberSumLeqMatch(n, m int) *NumberSumLeqMatch {
	used := min(n, m-1)
	type1 := m - used - 1
	return &NumberSumLeqMatch{n: n, m: m, used: used, type1: type1}
}

func (nm *NumberSumLeqMatch) MaxMatching() int {
	return nm.used
}

// 如果x没有匹配，返回-1
func (nm *NumberSumLeqMatch) Partner(x int) int {
	if x > nm.used {
		return -1
	}
	if x <= nm.type1 {
		return x
	}
	return nm.m - x
}

// 如果从数组1中删除a和从数组2中删除b，最大匹配数是多少.
func (nm *NumberSumLeqMatch) MaxMatching2(a, b int) int {
	if a >= nm.m && b >= nm.m {
		return nm.used
	}
	if a+b < nm.m && nm.type1 <= 1 {
		return nm.used - 2
	}
	return nm.used - 1
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
